### PR TITLE
Mailbox 279: JPAMappersTest should work with MESSAGE

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 1]
```diff
@@ -191,12 +136,22 @@ public long countUnseenMessagesInMailbox(Mailbox mailbox) throws MailboxExceptio
...
-            getEntityManager().remove(message);
+            AbstractJPAMailboxMessage jpaMessage = getEntityManager().find(AbstractJPAMailboxMessage.class, buildKey(mailbox, message));
+            getEntityManager().remove(jpaMessage);
+
         } catch (PersistenceException e) {
             throw new MailboxException("Delete of message " + message + " failed in mailbox " + mailbox, e);
         }
     }
 
+    private AbstractJPAMailboxMessage.MailboxIdUidKey buildKey(Mailbox mailbox, MailboxMessage message) {
```
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: What is this AbstractJpaMailboxMessage thing ?

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 2]
```diff
@@ -285,15 +240,71 @@ public MessageMetaData move(Mailbox mailbox, MailboxMessage original) throws Mai
...
+    @Override
+    public long getHighestModSeq(Mailbox mailbox) throws MailboxException {
+        return messageMetadataMapper.getHighestModSeq(mailbox);
+    }
+
+    private MessageMetaData copy(Mailbox mailbox, MessageUid uid, long modSeq, MailboxMessage original)
             throws MailboxException {
         MailboxMessage copy;
+        JPAMailbox currentMailbox;
+        if (mailbox instanceof JPAMailbox) {
```
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: It's not the first time I see this pattern, the right thing to do is to implement a JPAMailbox.from(Mailbox) that does that "optimization"

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 3]
```diff
@@ -285,15 +240,71 @@ public MessageMetaData move(Mailbox mailbox, MailboxMessage original) throws Mai
...
     @Override
-    protected MessageMetaData copy(Mailbox mailbox, MessageUid uid, long modSeq, MailboxMessage original)
+    public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        messageMetadataMapper.enrichMessage(mailbox, message);
+
+        return save(mailbox, message);
+    }
+
+    @Override
+    public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator,
```
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: the inheritance change brings this copy/paste if I understand correctly.
That's the trap of using inheritance where we should use composition.
Please go one step further by extracting pieces of code into classes that we can compose.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageUid;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+
+import com.google.common.base.Optional;
+
+public class MessageMetadata {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java]: This name makes me think it's an "object" or a "bean" but I don't expect it to call services like it does. So maybe the name is not ok ? or is it the implementation ?

**Comment by quynhn** [comment]
test this please

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/JPAMapperProvider.java [hunk 5]
```diff
@@ -124,6 +134,6 @@ private EntityManagerFactory createEntityManagerFactory() {
 
     @Override
     public List<Capabilities> getNotImplemented() {
```
Comment (mbaechler) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/JPAMapperProvider.java]: That's funny, usually we do the opposite for a good reason : when we add a capability to the system, all classes that don't implement it yet don't suddenly break.

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/TransactionalMessageMapper.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Throwables;
+
+public class TransactionalMessageMapper implements MessageMapper {
+    private final JPAMessageMapper wrapped;
+
+    public TransactionalMessageMapper(JPAMessageMapper wrapped) {
+        this.wrapped = wrapped;
```
Comment (mbaechler) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/TransactionalMessageMapper.java]: it's a delegate

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 7]
```diff
@@ -285,15 +240,71 @@ public MessageMetaData move(Mailbox mailbox, MailboxMessage original) throws Mai
...
+    private MessageMetaData copy(Mailbox mailbox, MessageUid uid, long modSeq, MailboxMessage original)
             throws MailboxException {
         MailboxMessage copy;
+        JPAMailbox currentMailbox;
+        if (mailbox instanceof JPAMailbox) {
+            currentMailbox = (JPAMailbox) mailbox;
+        } else {
+            currentMailbox = new JPAMailbox(mailbox);
+        }
         if (original instanceof JPAStreamingMailboxMessage) {
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: What you need here is 3 JPAMessageMappers : 
 - One for Streaming
 - One for Encrypted
 - One for the default

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    public MessageUid nextUid(Mailbox mailbox) throws MailboxException {
+        return uidProvider.nextUid(mailboxSession, mailbox);
+    }
+
+    public long nextModSeq(Mailbox mailbox) throws MailboxException {
+        if (modSeqProvider != null) {
+            return modSeqProvider.nextModSeq(mailboxSession, mailbox);
+        }
+        return -1;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java]: Why don't we fail here ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public void enrichMessage(Mailbox mailbox, MailboxMessage message) throws MailboxException { 
+        message.setUid(nextUid(mailbox));
+        
+        if (modSeqProvider != null) {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java]: Should never happen

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java (hunk 1 above)
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: The AbstractJpaMailboxMessage is the base Message model which contain all needed properties but base on advanced feature (Streaming or Encryption) we can support message which has header and body with stream or encrypted.
But for now, with guice, we just support default message. So do we need to keep them all as they are or remove all of them, remove abstract, just keep default model?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java (hunk 4 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java]: It's not a bean, it likes the MessageUtils so I changed to MessageUtils and move related functions to.

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java (hunk 3 above)
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: Yes, I move to MessageMetadata (later on, changed to MessageUtils)

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java (hunk 8 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java]: Cause modSeq never be null. Anyway, I added more Preconditions

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java (hunk 9 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageMetadata.java]: Yes

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java (hunk 7 above)
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: It depends on how can the DB can support, does it support streaming or encrypted data then system will support more advance feature (header and body can be streaming or can be encrypted). But for now, with guice, I've just seen default feature (no streaming and no encryption). So I don't know do we keep those or remove them all.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java [hunk 10]
```diff
@@ -178,6 +179,24 @@ public void messagesRetrievedUsingFetchTypeFullShouldHaveAttachmentsEmptyWhenNoA
         assertThat(retrievedMessageIterator.next().getAttachments()).isEmpty();
     }
     
+    @ContractTest
+    public void messagesCanBeRetrievedInMailboxWithRangeTypeOne() throws MailboxException, IOException{
+        saveMessages();
+        MessageMapper.FetchType fetchType = MessageMapper.FetchType.Full;
+        int limit =10;
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java]: missing space before 10

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java [hunk 11]
```diff
@@ -1,4 +1,4 @@
-/****************************************************************
+    /****************************************************************
```
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java]: Oops

Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageWithAttachmentMapperTest.java]: :+1: I rolled it back

**Comment by rouazana** [comment]
test this please

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java [hunk 12]
```diff
@@ -134,8 +134,8 @@ public AnnotationMapper createAnnotationMapper() throws MailboxException {
     }
 
     @Override
-    public List<Capabilities> getNotImplemented() {
-        return ImmutableList.of();
+    public List<Capabilities> getSupportedCapabilities() {
+        return Lists.newArrayList(Capabilities.values());
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java]: Why couldn't it be ImmutableList ?

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java [hunk 13]
```diff
@@ -96,6 +96,13 @@
     @Column(name = "MAILBOX_HIGHEST_MODSEQ", nullable = true)
     private long highestModSeq;
     
+    public static JPAMailbox from(Mailbox mailbox) {
+        if (mailbox instanceof JPAMailbox) {
+            return (JPAMailbox)mailbox;
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java]: You are missing a space : 

```
return (JPAMailbox) mailbox;
```

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMapperProvider.java [hunk 14]
```diff
@@ -20,7 +20,7 @@
 import org.apache.james.mailbox.store.mail.model.MapperProvider;
 import org.apache.james.mailbox.store.mail.model.MessageUidProvider;
 
-import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
```
Comment (chibenwa) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMapperProvider.java]: Why couldn't it be immutableList ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageUtils.java [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        return new MessageChangedFlags(updatedFlags.build().iterator(), changedFlags.build());
+    }
+
+    
+    public class MessageChangedFlags {
+        private final Iterator<UpdatedFlags> updatedFlags;
+        private final List<MailboxMessage> changedFlags;
+
+        public MessageChangedFlags(Iterator<UpdatedFlags> updatedFlags, List<MailboxMessage> changedFlags) {
+            super();
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageUtils.java]:
```
super();
```

Does not seem to be needed ...

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessageAssertTest.java [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+    public void setUp() {
+        benwaInboxMailbox = createMailbox(new MailboxPath("#private", "user", "INBOX"));
+
+        message1 = createMessage(benwaInboxMailbox, MESSAGE_ID, BODY_CONTENT1, BODY_START, new PropertyBuilder());
+        message2 = createMessage(benwaInboxMailbox, MESSAGE_ID, BODY_CONTENT2, BODY_START, new PropertyBuilder());
+    }
+
+    @Test
+    public void containsOnlyShouldWork() throws IOException {
+        actual = ImmutableList.of(message1, message2);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessageAssertTest.java]: actual should be a local variable here and for the one below

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+import java.util.List;
+
+import com.google.common.base.Function;
+import com.google.common.base.Objects;
+import com.google.common.collect.FluentIterable;
+
+public class ListMessagePropertiesAssert {
+    private final List<InnerProperty> propertyToInnerProperty(List<Property> properties) {
+        return FluentIterable.from(properties).transform(new Function<Property, InnerProperty>() {
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java]: break line between from and transform ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        assertThat(propertyToInnerProperty(actual)).containsOnlyElementsOf(propertyToInnerProperty(expected));
+    }
+    
+    private final class InnerProperty {
+        private final String namespace;
+        private final String name;
+        private final String value;
+
+        public InnerProperty(String namespace, String name, String value) {
+            super();
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java]: super not needed ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java (hunk 12 above)
Comment (quynhn) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProvider.java]: Yes, I changed to "ImmutableList.copyOf"

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java (hunk 13 above)
Comment (quynhn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAMailbox.java]: Done

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMapperProvider.java (hunk 14 above)
Comment (quynhn) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMapperProvider.java]: Yes, it's changed to ImmutableList.copyOf

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageUtils.java (hunk 15 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/MessageUtils.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessageAssertTest.java (hunk 16 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessageAssertTest.java]: OK, it's done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java (hunk 17 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java]: Yes

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java (hunk 18 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/ListMessagePropertiesAssert.java]: OK, it's removed

**Pull request merged**